### PR TITLE
Optimise picturefill for use on the guardian

### PR DESCRIPTION
**Pull request by sndrs**
Removes support for things we don't use, massively reduces DOM trashing and batches writes with `fastdom`:
- now a full AMD module
- returns method to call, rather than running simply by being present
- assumes DOM is complete, and so doesn't call itself in a `setInterval` loop
- uses fastdom
- removes `picture` support, since we're not using it
- removes type support checking
- removes width/css unit checking, since we always use `px`
  - for a page of 50 images, [this code](https://github.com/guardian/picturefill/blob/20d426c2f69b1f62651e6c27aba14f5ad47843f3/src/picturefill.js#L88) was being run 150 times, triggering layout 5 times each time
- removes `backfaceVisibilityFix`
  - we don't replace bitmaps with SVGs, and layout was triggered 3 times per image
- removes mixed content check

Before, Safari 8:

screen shot 2015-02-23 at 14 03 57

After:

screen shot 2015-02-23 at 14 09 51

cc @samdesborough @phamann @robertberry @gklopper: I _think_ we don't need any of these currently - look right to you?

### CODE REVIEW

**Comment by robertberry-zz** [comment]
```
removes width/css unit checking, since we always use px
```

We use % too (for full-media images).

**Comment by robertberry-zz** [comment]
This is amazing!

File: dist/picturefill.js [hunk 1]
```diff
@@ -47,678 +47,705 @@ window.matchMedia || (window.matchMedia = function() {
...
+		// 		style.zoom = currentZoom;
+		// 	}
+		// };
 
-			bestCandidate.url = pf.makeUrl( bestCandidate.url );
+		pf.setIntrinsicSize = (function() {
+			var urlCache = {};
+			var setSize = function( picImg, width, res ) {
+				fastdom.write( function() {
+					picImg.setAttribute( "width", parseInt(width / res, 10) );
```
Comment (robertberry-zz) [inline dist/picturefill.js]: I've already removed this from the `frontend` version, as it's buggy and sets incorrect widths.

We also don't need it to set the width. It's weird and overrides CSS. Why does it do it to begin with?!

**Comment by robertberry-zz** [comment]
Looks good to me. Have you tested it on frontend?

**Comment by sndrs** [comment]
> We use % too (for full-media images).

ah ok, i'll restore some of that unit checking stuff then.

**Comment by sndrs** [comment]
> I've already removed this from the frontend version, as it's buggy and sets incorrect widths.
> 
> We also don't need it to set the width. It's weird and overrides CSS. Why does it do it to begin with?!

yeah good point, that should go too

**Comment by sndrs** [comment]
> Have you tested it on frontend?

yeah i've been doing a fiddly dance of making the changes in frontend, and copy them over to this fork. is these get the thumbs up, I'm not sure what the value of maintaining it as a fork is (i.e. just have our own file in frontend)

**Comment by phamann** [comment]
>  I'm not sure what the value of maintaining it as a fork is

We did this so that we could pull in up-stream changes more easily, though may not be possibel after this PR. But still think it is useful to keep seperate. 

As discussed with @robertberry and @samdesborough the Picturefill team are having essentially a complete re-write for version 3 to address most of the issues we have found. So think we should steam ahead with this and check out version 3 when they are ready.

**Comment by robertberry-zz** [comment]
Yupyup. @sndrs Can you make a pull request to frontend? We'll do some testing here and see if we can merge this!

**Comment by sndrs** [comment]
> As discussed with @robertberry and @samdesborough the Picturefill team are having essentially a complete re-write for version 3 to address most of the issues we have found. So think we should steam ahead with this and check out version 3 when they are ready.

ok that's perfect then. 

> Can you make a pull request to frontend? We'll do some testing here and see if we can merge this!

yeah will do. going to look at that `setIntrinsicSize` method to check it's really not doign anything like it seems then a frontend PR.

**Comment by sndrs** [comment]
> We did this so that we could pull in up-stream changes more easily, though may not be possibel after this PR. But still think it is useful to keep seperate.

yeah, it was for that reason it's just commented out (hopefully eeasier to pull in changes). but if v3 is good then that will be ideal.

one thing to consider then though - alot (@robertberry) of this was for `picture` and edge cases we know we don't need. might still be worth maintaining a fork nonetheless. need to see though.

**Comment by phamann** [comment]
@sndrs re intrinsic sizing, see this open PR in Picutrefill's repo: https://github.com/scottjehl/picturefill/pull/442/files

**Comment by sndrs** [comment]
> re intrinsic sizing, see this open PR in Picutrefill's repo

I can't actually tell what it does - does it just cache sizes for subsequent runs of picturefill (e.g. on resize) that we now never run? disabling seems to have no effect at all...

**Comment by sndrs** [comment]
@robertberry 

> We use % too (for full-media images).

don't we still provide pixel values in `sizes` attribute though?

**Comment by robertberry-zz** [comment]
@sndrs No, 'cause on mobile breakpoints it is 100% of the viewport. Look at 'The Big Picture' at the bottom of http://www.theguardian.com/uk

**Comment by sndrs** [comment]
@robertberry ah ha, thanks

**Comment by gklopper** [comment]
I like

**Comment by samdesborough** [comment]
if including the unit check makes a big perf hit we should be able to add some extra breakpoints/pixel widths for that type of image

**Pull request merged**